plugins/wasi_nn/mlx.cpp

```diff
@@ -0,0 +1,344 @@
+    TokenizerPath = TokenizerPathView;
+  } else {
+    spdlog::error(
+        "[WASI-NN] MLX backend: Unable to retrieve the tokenizer option."sv);
+    Env.NNGraph.pop_back();
+    return ErrNo::InvalidArgument;
+  }
+  if (Doc.at_key("max_token").error() == simdjson::SUCCESS) {
+    uint64_t MaxToken;
+    auto Err = Doc["max_token"].get<uint64_t>().get(MaxToken);
+    if (Err) {
+      spdlog::error(
+          "[WASI-NN] MLX backend: Unable to retrieve the max_token option."sv);
+      Env.NNGraph.pop_back();
+      return ErrNo::InvalidArgument;
+    }
+    GraphRef.MaxToken = MaxToken;
+  }
+  if (Doc.at_key("q_bits").error() == simdjson::SUCCESS &&
+      Doc.at_key("group_size").error() == simdjson::SUCCESS &&
+      Doc.at_key("is_quantized").error() == simdjson::SUCCESS) {
+    uint64_t QBits;
+    uint64_t GroupSize;
+    bool IsQuantized;
+    auto ErrQBits = Doc["q_bits"].get<uint64_t>().get(QBits);
+    auto ErrGroupSize = Doc["group_size"].get<uint64_t>().get(GroupSize);
+    auto ErrIsQuantized = Doc["is_quantized"].get<bool>().get(IsQuantized);
+    if (ErrQBits || ErrGroupSize || ErrIsQuantized) {
+      spdlog::error(
+          "[WASI-NN] MLX backend: Unable to retrieve the q_bits or group_size option."sv);
+      Env.NNGraph.pop_back();
+      return ErrNo::InvalidArgument;
+    }
+    GraphRef.IsQuantized = IsQuantized;
+    GraphRef.QBits = QBits;
+    GraphRef.GroupSize = GroupSize;
+  }
+
+  // Load tokenizer.
+  if (!TokenizerPath.empty()) {
+    GraphRef.Tok =
+        tokenizers::Tokenizer::FromBlobJSON(loadBytesFromFile(TokenizerPath));
+  } else {
+    spdlog::error("[WASI-NN] MLX backend: Tokenizer path not found."sv);
+    Env.NNGraph.pop_back();
+    return ErrNo::InvalidArgument;
+  }
+
+  // Create Model.
+  if (GraphRef.ModelType == "tiny_llama_1.1B_chat_v1.0") {
+    GraphRef.Model = tinyLlama11BChatV10();
+    GraphRef.Prmopt = TinyLLaMAPrompt();
+  } else if (GraphRef.ModelType == "llama_3_8b") {
+    GraphRef.Model = llama38b();
+    GraphRef.Prmopt = LLaMA3Prompt();
+  } else if (GraphRef.ModelType == "llama_2_7b_chat_hf") {
+    GraphRef.Model = llama27bChat();
+    GraphRef.Prmopt = LLaMA2Prompt();
+  } else {
+    spdlog::error("[WASI-NN] MLX backend: Model type not supported."sv);
+    Env.NNGraph.pop_back();
+    return ErrNo::InvalidArgument;
+  }
+
+  if (GraphRef.QBits != 0 && GraphRef.GroupSize != 0 && GraphRef.IsQuantized) {
+    GraphRef.Model->toQuantized(GraphRef.GroupSize, GraphRef.QBits);
+  }
+
+  // Handle the model path.
+  for (size_t Idx = 0; Idx < Builders.size() - 1; Idx++) {
+    auto Weight = Builders[Idx];
+    const std::string BinModel(reinterpret_cast<char *>(Weight.data()),
+                               Weight.size());
+    spdlog::info("[WASI-NN] MLX BinModel: {}"sv, BinModel.size());
+    if (BinModel.size() == 0) {
+      Env.NNGraph.pop_back();
+      return ErrNo::InvalidArgument;
+    }
+    std::string ModelFilePath;
+    if (BinModel.substr(0, 8) == "preload:"sv) {
```

Comment:
If the model does not use `nn-preload,` we will no longer accept it. Most LLM models are larger than the Wasm memory limitation (4GB), and it's not possible to pass these kinds of models via arguments. We can show the error message to ask users to pass the model through the `nn-preload` here.

Reply:
Never mind, let's ignore this.